app/parse-payload.ts

```diff
@@ -0,0 +1,360 @@
+export function parsePayload(payload: string) {
+  const rows = payload.split("\n");
+
+  const response: _Response = {
+    _chunks: new Map(),
+    _fromJSON: undefined as any,
+  };
+
+  response._fromJSON = createFromJSONCallback(response as _Response);
+
+  const parsedRows = rows
+    .map((row) => {
+      const processed = processFullRow(response, row);
+      if (!processed) {
+        return null;
+      }
+      response._chunks.set(processed.id, processed.row);
+      return { raw: row, parsed: processed };
+    })
+    .filter((x) => !!x)
+    .map((x) => x!);
+
+  return {
+    raw: payload,
+    rows: parsedRows,
+    chunks: response._chunks,
+  };
+}
+
+export type ParsedPayload = ReturnType<typeof parsePayload>;
+export type Row = ParsedPayload["rows"][number];
+
+type _Response = {
```

Comment:
this is a bit weird because it's ported almost directly from the react code, but doesn't quite make sense for us here. not sure if we wanna
- keep it aligned for easier sync of upstream updates?
- OR clean it up to be more appropriate for our usecase?

Reply:
I think we should start with keeping it aligned but we should talk with @sebmarkbage